File: JavaScript/26_Design/LC_604_Design_Compressed_String_Iterator.js
/**
 * @param {string} compressedString
 */

var StringIterator = function(compressedString) {
    this.index = 0;
    this.group = Array.from(compressedString.match(/\D|\d+/g))
};

/**
 * @return {character}
 */
StringIterator.prototype.next = function() {
    if(!this.hasNext()){
        return ' ';
    }

    const res = this.group[this.index];
    this.group[this.index + 1]--;

    if(this.group[this.index + 1] === 0){
        this.index += 2;
    }

    return res;
};

/**
 * @return {boolean}
 */
StringIterator.prototype.hasNext = function() {
    return this.group[this.index + 1] > 0;
};

/**
 * Your StringIterator object will be instantiated and called as such:
 * var obj = new StringIterator(compressedString)
 * var param_1 = obj.next()
 * var param_2 = obj.hasNext()
 */

File: Java/26_Design/LC_604_DesignCompressedStringIterator.java
class StringIterator {
    String res;
    int ptr = 0;
    int num = 0;
    char ch = ' ';

    public StringIterator(String s) {
        res = s;
    }

    // 每次取下一个值的时候，对缓存 ch 和 num 进行计算
    public char next() {
        if (!hasNext()) {
            return ' ';
        }
        if (num == 0) {
            ch = res.charAt(ptr++);
            while (ptr < res.length() && Character.isDigit(res.charAt(ptr))) {
                num = num * 10 + res.charAt(ptr) - '0';
                ptr++;
            }
        }
        num--;
        return ch;
    }

    public boolean hasNext() {
        return ptr != res.length() || num != 0;
    }
}

public class LC_604_DesignCompressedStringIterator {
    
}